docs/paper/bitvm2_fast_recap.tex

```diff
@@ -135,71 +171,245 @@ \subsection{\texttt{DispoveScript}}\label{sec:dispove-script}
+      $(e_1,\dots,e_k) \gets \mathsf{Enc}(m)$ and then check whether:
+      \begin{equation*}
+        H^{(d-e_j)}(\sigma_j) = y_j, \quad j \in \{1,\dots,k\}.
+      \end{equation*}
+  \end{itemize}
+\end{definition}
+
+That being said, by taking the intermediate states
+${\{z_j\}}_{1 \leq j \leq n}$ as the message for the Winternitz signature, we
+form one-time key pairs
+${\{(\mathsf{sk}_j,\mathsf{pk}_j)\}}_{1 \leq j \leq n}$ and signatures
+${\{\sigma_j\}}_{1 \leq j \leq n}$, respectively (where each of
+  $\mathsf{pk}_j$, $\mathsf{sk}_j$, and $\sigma_j$ corresponds to the
+intermediate variable $z_j$). Then, to spend the output from the
+\texttt{Assert} transaction using the $\texttt{DisproveScript}_j$
+script, the challenger is required to add the corresponding states
+$z_j$, $z_{j-1}$, and corresponding signatures $\sigma_j$,
+$\sigma_{j-1}$ to the stack in the \texttt{scriptSig}, making the
+\texttt{scriptSig} of the transaction input like this:
+
+\begin{empheqboxed}
+  \begin{align*}
+    &\elem{z_{j-1}} \opcode{OP\_DUP} \elem{\sigma_{j-1}}
+    \elem{\mathsf{pk}_{j-1}} \opcode{OP\_WINTERNITZVERIFY} \\
+    &\elem{z_{j}} \opcode{OP\_DUP} \elem{\sigma_{j}}
+    \elem{\mathsf{pk}_{j}} \opcode{OP\_WINTERNITZVERIFY} \\
+    &\elem{f_j} \opcode{OP\_EQUAL} \opcode{OP\_NOT}
+  \end{align*}
+\end{empheqboxed}
+
+where \texttt{OP\_WINTERNITZVERIFY} is the verification of the
+Winternitz signature (commitment), described in Bitcoin Script (as
+  Bitcoin Script does not have a built-in \texttt{OP\_CODE} for
+Winternitz signatures)\footnote{Its implementation can be found here:
+\url{https://github.com/distributed-lab/bitvm2-splitter/blob/feature/winternitz/bitcoin-winternitz/src/lib.rs}.}.
+
+The correct implementation of \texttt{OP\_WINTERNITZVERIFY} in Bitcoin
+Script would require the $\mathsf{Enc}(m)$ to split the state into $d$
+digit numbers or recover the state from $d$ digit numbers on stack,
+which without the bitwise operations or \texttt{OP\_CAT} for larger
+than 32-bit values is nearly impossible.
+
+\subsubsection{32-bit stack elements}\label{sec:winternitz-31-bit-integers}
+
+Instead of bitwise operations, script can use arithmetic ones, but the
+Bitcoin Script supports opcodes such as \texttt{OP\_ADD}, only for
+32-bit signed integers, and as the last one is reserved for a sign,
+only 31 bits can be used to store the state. This could be considered
+as limitation, but as most of the complex arithmetic operations in
+BitVM are done using only 32-bit integers, we can easily fix $\ell$ as
+$\ell = 32$ for all intermidiate states.
+
+By defining the $\mathsf{Enc}(m)$ as a domination free function
+$P(m)$, like described in~\cite{applied-crypto}, $d+1$ becomes the
+power of two which splits $m$ by some number of equal chunks of
+$\log_2{(d+1)}$ bits, and $k$ becomes the sum of $n_0$ --- the number of
+$d$-digit numbers from $m$, and $n_1$ a checksum (see
+table~\ref{tab:winternitz}).
+
+\iffalse{}
```

Comment:
This macro excludes python code below, but I found it useful for myself to keed it somewhere. Not sure should we move it somewhere or keep it here